### PR TITLE
[SPARK-17495] [SQL] Support Decimal type in Hive-hash

### PR DESCRIPTION
**Pull request by tejasapatil**
## What changes were proposed in this pull request?

Hive hash to support Decimal datatype. [Hive internally normalises decimals](https://github.com/apache/hive/blob/4ba713ccd85c3706d195aeef9476e6e6363f1c21/storage-api/src/java/org/apache/hadoop/hive/common/type/HiveDecimalV1.java#L307) and I have ported that logic as-is to HiveHash.

## How was this patch tested?

Added unit tests

### CODE REVIEW

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
Jenkins retest this please

last failure was weird termination of jenkins:

```
[info] NaiveBayesClusterSuite:
Traceback (most recent call last):
  File "./dev/run-tests-jenkins.py", line 226, in <module>
    main()
  File "./dev/run-tests-jenkins.py", line 213, in main
    test_result_code, test_result_note = run_tests(tests_timeout)
  File "./dev/run-tests-jenkins.py", line 140, in run_tests
    test_result_note = ' * This patch **fails %s**.' % failure_note_by_errcode[test_result_code]
KeyError: -9
[error] running /home/jenkins/workspace/SparkPullRequestBuilder@2/build/sbt -Phadoop-2.6 -Phive-thriftserver -Phive -Dtest.exclude.tags=org.apache.spark.tags.ExtendedHiveTest,org.apache.spark.tags.ExtendedYarnTest hive-thriftserver/test mllib/test hive/test examples/test sql/test sql-kafka-0-10/test catalyst/test ; process was terminated by signal 9
```

**Comment by tejasapatil** [comment]
retest this please

**Comment by tejasapatil** [comment]
@cloud-fan @gatorsmile : can you please review this PR ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 1]
```diff
@@ -371,6 +370,48 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+        precision: Int,
+        scale: Int,
+        expected: Long): Unit = {
+      val decimal = Decimal.apply(new java.math.BigDecimal(input))
+      decimal.changePrecision(precision, scale)
+      val decimalType = DataTypes.createDecimalType(precision, scale)
+      checkHiveHash(decimal, decimalType, expected)
+    }
+
+    checkHiveHashForDecimal("18", 38, 0, 558)
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: A quick question: these expected values are got from Hive?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 2]
```diff
@@ -371,6 +370,48 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+    checkHiveHashForDecimal("-18", 38, 0, -558)
+    checkHiveHashForDecimal("-18", 38, 12, -558)
+    checkHiveHashForDecimal("18446744073709001000", 38, 19, -17070057)
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]:
```
hive> select HASH(CAST("-18446744073709001000" AS DECIMAL(38,19)));
OK
0
Time taken: 0.035 seconds, Fetched: 1 row(s)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: I did a quick check. Most are right, but some of them do not match

Comment (tejasapatil) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: These were generated over Hive 1.2.1.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala (hunk 2 above)
Comment (tejasapatil) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: The main reason why not all of them match is because difference in how scale and precision are enforced within Hive vs Spark.

Hive does it using its own custom logic : https://github.com/apache/hive/blob/branch-1.2/common/src/java/org/apache/hadoop/hive/common/type/HiveDecimal.java#L274

Spark has its own way : https://github.com/apache/spark/blob/0e2405490f2056728d1353abbac6f3ea177ae533/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L230

Now when one does `CAST(-18446744073709001000BD AS DECIMAL(38,19))`, it does NOT fit in Hive's range and it will convert it to `null`... and `HASH()` over `null` will return 0.

In case of Spark, `CAST(-18446744073709001000BD AS DECIMAL(38,19))` is valid and running `HASH()` over it thus gives some non-zero result.
 
TLDR: this difference is before the hashing function comes into the picture. Making this in sync would mean the semantics of Decimal in Spark need to be matched with that in Hive. I don't think its a good idea to embark on that as it will be a breaking change plus this PR is not a strong reason to push for that. Hive-hash is best effort.

Comment (tejasapatil) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: I figured out the problem was with the test case being not looking at the result of `decimal.changePrecision`. Fixed.

**Comment by tejasapatil** [comment]
Jenkins test this please

**Comment by tejasapatil** [comment]
cc @cloud-fan @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 3]
```diff
@@ -635,6 +636,16 @@ case class HiveHash(children: Seq[Expression]) extends HashExpression[Int] {
   override protected def genHashBytes(b: String, result: String): String =
     s"$result = $hasherClassName.hashUnsafeBytes($b, Platform.BYTE_ARRAY_OFFSET, $b.length);"
 
+  override protected def genHashDecimal(
+      ctx: CodegenContext,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: where do we use `ctx` and `d`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 4]
```diff
@@ -635,6 +636,16 @@ case class HiveHash(children: Seq[Expression]) extends HashExpression[Int] {
...
+      d: DecimalType,
+      input: String,
+      result: String): String = {
+    s"""
+      $result = org.apache.spark.sql.catalyst.expressions.HiveHashFunction.normalizeDecimal(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: `${classOf[HiveHashFunction].getName}.normalizeDecimal`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 5]
```diff
@@ -371,6 +370,51 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+      val decimalType = DataTypes.createDecimalType(precision, scale)
+      val decimal = {
+        val value = Decimal.apply(new java.math.BigDecimal(input))
+        if (value.changePrecision(precision, scale)) value else null
+      }
+
+      checkHiveHash(decimal, decimalType, expected)
+    }
+
+    checkHiveHashForDecimal("18", 38, 0, 558)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: hmm it's hard to guarantee that we can produce same hash value as hive, can we run hive in the test and compare the result with spark?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 3 above)
Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: They both aren't used but are a part of the method signature since the default impl in abstract class needs those : https://github.com/apache/spark/blob/3e40f6c3d6fc0bcd828d09031fa3994925394889/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L321

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala (hunk 5 above)
Comment (tejasapatil) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala]: The expected values are generated using hive 1.2.1. My original approach was to depend on Hive for generating expected values but [as per discussion in a related PR](https://github.com/apache/spark/pull/15047#issuecomment-247231919), I was suggested to hardcode expected values

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 4 above)
Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: `HiveHashFunction` is an object so cannot do `classOf[]`. Tried `HiveHashFunction.getClass.getName` as per other places in the codebase

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 6]
```diff
@@ -732,6 +743,51 @@ object HiveHashFunction extends InterpretedHashFunction {
...
+      return null
+    }
+
+    def trimDecimal(input: BigDecimal) = {
+      var result = input
+      if (result.compareTo(BigDecimal.ZERO) == 0) {
+        // Special case for 0, because java doesn't strip zeros correctly on that number.
+        result = BigDecimal.ZERO
+      }
+      else {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: `} else {`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 7]
```diff
@@ -732,6 +743,51 @@ object HiveHashFunction extends InterpretedHashFunction {
...
   }
 
+  private val HiveDecimalMaxPrecision = 38
+  private val HiveDecimalMaxScale = 38
+
+  // Mimics normalization done for decimals in Hive at HiveDecimalV1.normalize()
+  def normalizeDecimal(input: BigDecimal, allowRounding: Boolean): BigDecimal = {
+    if (input == null) {
+      return null
+    }
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: Nit: `if (input == null) return null`

**Comment by gatorsmile** [comment]
Let me manually check whether the results are consistent with Hive.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 6 above)
Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: changed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 7 above)
Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: changed

**Comment by tejasapatil** [comment]
@gatorsmile : I really appreciate your help in reviewing this PR to the extent that you are manually checking the hashes over Hive. If you haven't already embarked on that, here is the set of hive queries corresponding to the test case in the PR which you can easily copy paste:

```
SELECT HASH(CAST(18BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(-18BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(-18BD AS DECIMAL(38, 12)));
SELECT HASH(CAST(18446744073709001000BD AS DECIMAL(38, 19)));
SELECT HASH(CAST(-18446744073709001000BD AS DECIMAL(38, 22)));
SELECT HASH(CAST(-18446744073709001000BD AS DECIMAL(38, 3)));
SELECT HASH(CAST(18446744073709001000BD AS DECIMAL(38, 4)));
SELECT HASH(CAST(9223372036854775807BD AS DECIMAL(38, 4)));
SELECT HASH(CAST(-9223372036854775807BD AS DECIMAL(38, 5)));
SELECT HASH(CAST(00000.00000000000BD AS DECIMAL(38, 34)));
SELECT HASH(CAST(-00000.00000000000BD AS DECIMAL(38, 11)));
SELECT HASH(CAST(123456.1234567890BD AS DECIMAL(38, 2)));
SELECT HASH(CAST(123456.1234567890BD AS DECIMAL(38, 20)));
SELECT HASH(CAST(123456.1234567890BD AS DECIMAL(38, 10)));
SELECT HASH(CAST(-123456.1234567890BD AS DECIMAL(38, 10)));
SELECT HASH(CAST(123456.1234567890BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(-123456.1234567890BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(123456.1234567890BD AS DECIMAL(38, 20)));
SELECT HASH(CAST(-123456.1234567890BD AS DECIMAL(38, 20)));
SELECT HASH(CAST(123456.123456789012345678901234567890BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(-123456.123456789012345678901234567890BD AS DECIMAL(38, 0)));
SELECT HASH(CAST(123456.123456789012345678901234567890BD AS DECIMAL(38, 10)));
SELECT HASH(CAST(-123456.123456789012345678901234567890BD AS DECIMAL(38, 10)));
SELECT HASH(CAST(123456.123456789012345678901234567890BD AS DECIMAL(38, 20)));
SELECT HASH(CAST(-123456.123456789012345678901234567890BD AS DECIMAL(38, 20)));
SELECT HASH(CAST(123456.123456789012345678901234567890BD AS DECIMAL(38, 30)));
SELECT HASH(CAST(-123456.123456789012345678901234567890BD AS DECIMAL(38, 30)));
SELECT HASH(CAST(123456.123456789012345678901234567890BD AS DECIMAL(38, 31)));
```

**Comment by gatorsmile** [comment]
Thank you! I checked Hive 2.1. It has the exactly same hash values.

**Comment by gatorsmile** [comment]
LGTM. cc @cloud-fan for final signing-off

**Comment by tejasapatil** [comment]
@cloud-fan ping !!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 8]
```diff
@@ -732,6 +743,48 @@ object HiveHashFunction extends InterpretedHashFunction {
     HiveHasher.hashUnsafeBytes(base, offset, len)
   }
 
+  private val HiveDecimalMaxPrecision = 38
+  private val HiveDecimalMaxScale = 38
+
+  // Mimics normalization done for decimals in Hive at HiveDecimalV1.normalize()
+  def normalizeDecimal(input: BigDecimal, allowRounding: Boolean): BigDecimal = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: allowRounding will never be false?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 9]
```diff
@@ -732,6 +743,48 @@ object HiveHashFunction extends InterpretedHashFunction {
     HiveHasher.hashUnsafeBytes(base, offset, len)
   }
 
+  private val HiveDecimalMaxPrecision = 38
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: nit: `HIVE_DECIMAL_MAX_PRECISION`

Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: renamed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 8 above)
Comment (tejasapatil) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: removed that param

**Comment by cloud-fan** [comment]
LGTM(if tests pass)

**Comment by cloud-fan** [comment]
thanks, merging to master!
